webapp/src/components/Breadcrumbs.tsx

```diff
@@ -0,0 +1,59 @@
+import { FC, useMemo } from 'react';
+import { Breadcrumbs as MuiBreadcrumbs, Link } from '@mui/material';
+
+type BreadcrumbsProps = {
+  languageName: string;
+  messageId?: string;
+  projectName: string;
+};
+
+type Breadcrumb = {
+  href: string;
+  last: boolean;
+  text: string;
+};
+
+const Breadcrumbs: FC<BreadcrumbsProps> = ({
+  languageName,
+  messageId,
+  projectName,
+}) => {
+  const breadcrumbs: Breadcrumb[] = useMemo(
+    () =>
+      (messageId?.split('.') || []).map((part, i) => {
+        const href = `/projects/${projectName}/${languageName}/${messageId!
+          .split('.')
+          .slice(0, i + 1)
+          .join('.')}`;
+        return {
+          href,
+          last: i === messageId!.split('.').length - 1,
+          text: part,
+        };
+      }, []),
+    [languageName, messageId, projectName],
+  );
+
+  return (
+    <MuiBreadcrumbs
+      aria-label="Breadcrumb navigation"
+      sx={{
+        maxWidth: '100%',
+        overflow: 'hidden',
+      }}
+    >
+      {breadcrumbs.map((breadcrumb) => (
+        <Link
+          key={breadcrumb.href}
+          color={breadcrumb.last ? 'text.primary' : 'text.secondary'}
+          href={breadcrumb.href}
```

Comment:
I don't think React has much XSS protection for href props. A user could be tricked into opening a lyra URL with something like `javascript:...` in the `messageId` route parameter. If the user then clicks that breadcrumb, the attacker gets control over the users' browser. I couldn't get around the browser URL-encoding necessary characters for this attack so maybe it's not possible with Next route parameters.

It would be safer if Lyra validated that the parameter matches messages before reflecting route parameters like this. It doesn't really matter much as long as we only have anonymous access anyway. An attacker can already do everything that a user can. The only benfit would be proxying through the user's IP addree. Let's save this for a separate issue.

Reply:
Loved this comment because it awoke an awareness in me of how much I've come to ignore these issues in my recent work. It's actually fun to think about too. Anyway, I fixed it in https://github.com/zetkin/lyra/pull/118/commits/8561d82b48e8c4a79a7e90ecf6c11dea1dd0b0ac because those requests should be 404ing anyway so it's relatively simple in this case at least.

| Before | After |
|-|-|
| ![Screenshot 2024-07-23 at 20 19 08](https://github.com/user-attachments/assets/f75c3ad1-062b-4b94-96d3-9eca4604c73b) | ![Screenshot 2024-07-23 at 20 18 58](https://github.com/user-attachments/assets/efd6e18e-e4db-4537-bb67-263b02d94c7d) |